File: src/main/java/org/lanternpowered/server/advancement/LanternDisplayInfoBuilder.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.advancement;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

import org.spongepowered.api.advancement.AdvancementType;
import org.spongepowered.api.advancement.AdvancementTypes;
import org.spongepowered.api.advancement.DisplayInfo;
import org.spongepowered.api.item.inventory.ItemStackSnapshot;
import org.spongepowered.api.text.Text;

@SuppressWarnings({"NullableProblems", "ConstantConditions"})
public final class LanternDisplayInfoBuilder implements DisplayInfo.Builder {

    AdvancementType advancementType;
    Text description;
    Text title;
    ItemStackSnapshot icon;
    boolean showToast;
    boolean announceToChat;
    boolean hidden;

    public LanternDisplayInfoBuilder() {
        reset();
    }

    @Override
    public DisplayInfo.Builder type(AdvancementType advancementType) {
        checkNotNull(advancementType, "advancementType");
        this.advancementType = advancementType;
        return this;
    }

    @Override
    public DisplayInfo.Builder description(Text description) {
        checkNotNull(description, "description");
        this.description = description;
        return this;
    }

    @Override
    public DisplayInfo.Builder title(Text title) {
        checkNotNull(title, "title");
        this.title = title;
        return this;
    }

    @Override
    public DisplayInfo.Builder icon(ItemStackSnapshot itemStackSnapshot) {
        checkNotNull(itemStackSnapshot, "itemStackSnapshot");
        this.icon = itemStackSnapshot;
        return this;
    }

    @Override
    public DisplayInfo.Builder showToast(boolean showToast) {
        this.showToast = showToast;
        return this;
    }

    @Override
    public DisplayInfo.Builder announceToChat(boolean announceToChat) {
        this.announceToChat = announceToChat;
        return this;
    }

    @Override
    public DisplayInfo.Builder hidden(boolean hidden) {
        this.hidden = hidden;
        return this;
    }

    @Override
    public DisplayInfo build() {
        checkState(this.title != null, "Title has not been set");
        checkState(this.icon != null, "Icon has not been set");
        return new LanternDisplayInfo(this);
    }

    @Override
    public DisplayInfo.Builder from(DisplayInfo value) {
        this.icon = value.getIcon();
        this.description = value.getDescription();
        this.advancementType = value.getType();
        this.announceToChat = value.doesAnnounceToChat();
        this.hidden = value.isHidden();
        this.showToast = value.doesShowToast();
        this.title = value.getTitle();
        return this;
    }

    @Override
    public DisplayInfo.Builder reset() {
        this.icon = null;
        this.description = Text.empty();
        this.advancementType = AdvancementTypes.TASK;
        this.announceToChat = true;
        this.hidden = false;
        this.showToast = true;
        this.title = null;
        return this;
    }
}

File: src/main/kotlin/org/lanternpowered/server/advancement/LanternDisplayInfoBuilder.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.advancement

import org.spongepowered.api.advancement.AdvancementType
import org.spongepowered.api.advancement.AdvancementTypes
import org.spongepowered.api.advancement.DisplayInfo
import org.spongepowered.api.item.inventory.ItemStackSnapshot
import org.spongepowered.api.text.Text

class LanternDisplayInfoBuilder : DisplayInfo.Builder {

    private var type: AdvancementType = AdvancementTypes.TASK.get()
    private var description: Text = Text.empty()
    private var title: Text? = null
    private var icon: ItemStackSnapshot? = null
    private var showToast = false
    private var announceToChat = false
    private var hidden = false

    override fun type(advancementType: AdvancementType): DisplayInfo.Builder = apply { this.type = advancementType }
    override fun description(description: Text): DisplayInfo.Builder = apply { this.description = description }
    override fun title(title: Text): DisplayInfo.Builder = apply { this.title = title }
    override fun icon(itemStackSnapshot: ItemStackSnapshot): DisplayInfo.Builder = apply { this.icon = itemStackSnapshot }
    override fun showToast(showToast: Boolean): DisplayInfo.Builder = apply { this.showToast = showToast }
    override fun announceToChat(announceToChat: Boolean): DisplayInfo.Builder = apply { this.announceToChat = announceToChat }
    override fun hidden(hidden: Boolean): DisplayInfo.Builder = apply { this.hidden = hidden }

    override fun build(): DisplayInfo {
        val title = checkNotNull(this.title) { "Title has not been set" }
        val icon = checkNotNull(this.icon) { "Icon has not been set" }
        return LanternDisplayInfo(title, this.description, icon, this.type,
                this.showToast, this.announceToChat, this.hidden)
    }

    override fun from(value: DisplayInfo): DisplayInfo.Builder = apply {
        this.icon = value.icon
        this.description = value.description
        this.type = value.type
        this.announceToChat = value.doesAnnounceToChat()
        this.hidden = value.isHidden
        this.showToast = value.doesShowToast()
        this.title = value.title
    }

    override fun reset(): DisplayInfo.Builder = apply {
        this.icon = null
        this.description = Text.empty()
        this.type = AdvancementTypes.TASK.get()
        this.announceToChat = true
        this.hidden = false
        this.showToast = true
        this.title = null
    }
}

private data class LanternDisplayInfo internal constructor(
        private val title: Text,
        private val description: Text,
        private val icon: ItemStackSnapshot,
        private val type: AdvancementType,
        private val showToast: Boolean,
        private val announceToChat: Boolean,
        private val hidden: Boolean
) : DisplayInfo {
    override fun getType(): AdvancementType = this.type
    override fun getDescription(): Text = this.description
    override fun getIcon(): ItemStackSnapshot = this.icon
    override fun getTitle(): Text = this.title
    override fun doesShowToast(): Boolean = this.showToast
    override fun doesAnnounceToChat(): Boolean = this.announceToChat
    override fun isHidden(): Boolean = this.hidden
}
